addons/account/models/account_move.py

```diff
@@ -1845,41 +1912,18 @@ def _check_unique_sequence_number(self):
 
     @api.constrains('ref', 'move_type', 'partner_id', 'journal_id', 'invoice_date', 'state')
     def _check_duplicate_supplier_reference(self):
-        moves = self.filtered(lambda move: move.state == 'posted' and move.is_purchase_document() and move.ref)
-        if not moves:
-            return
-
-        self.env["account.move"].flush([
-            "ref", "move_type", "invoice_date", "journal_id",
-            "company_id", "partner_id", "commercial_partner_id",
-        ])
-        self.env["account.journal"].flush(["company_id"])
-        self.env["res.partner"].flush(["commercial_partner_id"])
-
-        # /!\ Computed stored fields are not yet inside the database.
-        self._cr.execute('''
-            SELECT move2.id
-            FROM account_move move
-            JOIN account_journal journal ON journal.id = move.journal_id
-            JOIN res_partner partner ON partner.id = move.partner_id
-            INNER JOIN account_move move2 ON
-                move2.ref = move.ref
-                AND move2.company_id = journal.company_id
-                AND move2.commercial_partner_id = partner.commercial_partner_id
-                AND move2.move_type = move.move_type
-                AND (move.invoice_date is NULL OR move2.invoice_date = move.invoice_date)
-                AND move2.id != move.id
-            WHERE move.id IN %s
-        ''', [tuple(moves.ids)])
-        duplicated_moves = self.browse([r[0] for r in self._cr.fetchall()])
-        if duplicated_moves:
-            raise ValidationError(_('Duplicated vendor reference detected. You probably encoded twice the same vendor bill/credit note:\n%s') % "\n".join(
-                duplicated_moves.mapped(lambda m: "%(partner)s - %(ref)s - %(date)s" % {
-                    'ref': m.ref,
-                    'partner': m.partner_id.display_name,
-                    'date': format_date(self.env, m.invoice_date),
-                })
-            ))
+        """ Assert the move which is about to be posted isn't a duplicated move from another posted entry"""
+        move_to_duplicate_moves = self.filtered(lambda m: m.state == 'posted')._fetch_duplicate_supplier_reference(only_posted=True)
+        if any(m for m in move_to_duplicate_moves.values()):
+            duplicated_moves = sum(move_to_duplicate_moves.values(), self.env['account.move'])
```

Comment:
Wouldn't it make sense to have the keys of the dict as `<ref, date>` and the values are the ids? And basically just "pretty print" the dict in the message?

Reply:
Do you mean creating a new dict in this method and print it or do you mean even do that directly in the SQL query by grouping by `ref, date` ?

Reply:
That would be the return value of `_fetch_duplicate_supplier_reference`

Reply:
I don't like it.

- Either I go for a raw pretty print of the dict, I will show them ids of invoice which are meaningless to lambda user.
- Either I need to post_process the dict datas too which seems to me the same as the actual code.

And I think that even if I go with option 1 to make my life easier, it makes the `_fetch_duplicate_supplier_reference` and the `_compute_duplicated_ref_ids` method a little bit more complicated which cancels the benefits of making my life easier.

Also, I don't want to go with option 1 because ids are meaningless to lambda user. 
For example, I have an ex-colleague of mine working from time to time with Odoo Accounting. I think she would be mad at me if I tell her that I could have print the sequence number of the invoice but that I chose to print ids.

Reply:
Maybe we could just put a redirect warning and send the user to the list of duplicate.
But that could be messy if they get a lot of duplicate due to an EDI